contracts/CollectiveApprovalMechanism.cdc

```diff
@@ -0,0 +1,312 @@
+// CollectiveApprovalMechanism (aka CAM) is a contract to bring multi-sig management
+// and execution of work on-chain instead of using flow's built-in key-based approach
+access(all) contract CollectiveApprovalMechanism {
+    // Allows participating in a vote, if a Voter resource owner is marked as an owner
+    access(all) entitlement Vote
+    // Allows proposing a new executable to a manager from a Voter resource
+    access(all) entitlement Propose
+
+    access(all) let ManagerStoragePath: StoragePath
+    access(all) let VoterStoragePath: StoragePath
+
+    // Universal events, these are executed in the lifecycle of a Manager/Proposal/Vote regardless of the executable
+    access(all) event ManagerCreated(managerAddress: Address, uuid: UInt64, voters: {Address: UFix64})
+    access(all) event ProposalAdded(uuid: UInt64, title: String, description: String, executableType: String)
+    access(all) event VoteCast(managerAddress: Address, proposalId: UInt64, voterAddress: Address, weight: UFix64, approved: Bool)
+    access(all) event ProposalRun(managerAddress: Address, proposalId: UInt64, approvals: {Address: UFix64}, rejections: {Address: UFix64}, title: String, description: String)
+
+    // Executable-specific events. These are related to the built-in definitions of Executables that are needed to bootstrap
+    // newly made managers
+    access(all) event ChangeApprovedTypeExecutableCreated(uuid: UInt64, executableType: String, approved: Bool)
+    access(all) event ChangeApprovedTypeExecutableRun(uuid: UInt64, executableType: String, approved: Bool)
+    access(all) event ChangeVotersExecutableCreated(uuid: UInt64, voters: {Address: UFix64})
+    access(all) event ChangeVotersExecutableRun(uuid: UInt64, voters: {Address: UFix64})
+
+    // The building block of arbitrary code execution. If enough votes are cast in favor of
+    // a proposal, then its executable is able to be run. When run, an executable will get
+    // the fully entitled account which corresponds which the manager is in control of.
+    access(all) resource interface Executable {
+        // Run the executable if it is able to be run
+        access(contract) fun run(acct: auth(Storage, Contracts, Keys, Inbox, Capabilities) &Account)
```

Comment:
the fact that the accounts that runs this now needs priviledges means that this cannot be run by any account?

Would it be worth looking at not sending in any account here? maybe the resources can have an init method to set different accounts they can be run upon?

Reply:
nvm, i mistead the way things are run.

Reply:
My thinking for why you need to supply account here is that it is basically akin to a transaction. If I want to, for instance, update a contract definition if I get enough approval votes, I need to be able to somehow access the account to do so.

If I need the account reference to make the executable, then I do not even need to get anyone's approval and would just skip that step altogether